### PR TITLE
add Tair as vector store

### PR DESCRIPTION
**Pull request by seth-hg**
Hi all,

This PR adds a TairVectorStore which is backed by the Tair in-memory database. Tair is a Redis-compatible cloud native database developed by Alibaba Cloud. It can also be [used as a vector database](https://www.alibabacloud.com/help/en/tair/latest/tairvector). I think this could be useful.

Example notebook and unit tests are also included.

### CODE REVIEW

**Review by Disiok — APPROVED** [review]
This looks great. 

One clarifying question. Even though we add the additional Node metadata into the DB, it doesn't look like there's a way to do proper metadata filtering, as well as retrieving them right now. Is this understanding correct?

**Comment by seth-hg** [comment]
> This looks great.
> 
> One clarifying question. Even though we add the additional Node metadata into the DB, it doesn't look like there's a way to do proper metadata filtering, as well as retrieving them right now. Is this understanding correct?

Actually, Tair can do metadata filtering. I will try to figure out how to convert MetadataFilters into Tair's filter expression.

**Comment by Disiok** [comment]
> > This looks great.
> > One clarifying question. Even though we add the additional Node metadata into the DB, it doesn't look like there's a way to do proper metadata filtering, as well as retrieving them right now. Is this understanding correct?
> 
> Actually, Tair can do metadata filtering. I will try to figure out how to convert MetadataFilters into Tair's filter expression.

Great! This is up to you, I can help land this as is, and have a follow up to update, or wait for you to update it and merge later!

File: llama_index/vector_stores/tair.py [hunk 1]
```diff
@@ -140,8 +150,8 @@ def add(self, embedding_results: List[NodeWithEmbedding]) -> List[str]:
                 "doc_id": result.ref_doc_id,
                 "text": result.node.get_text(),
             }
-            additional_metadata = node_to_metadata_dict(result.node)
-            attributes.update(additional_metadata)
+            if result.node.node_info is not None:
+                attributes.update(result.node.node_info)
```
Comment (Disiok) [inline llama_index/vector_stores/tair.py:154]: nit: could we use the utility in `llama_index/vector_stores/utils.py` to get metadata dict from the Node (and re-creating Node from metadata dict)?

Comment (Disiok) [reply llama_index/vector_stores/tair.py:154]: You can take a look at `llama_index/vector_stores/chroma.py` for reference.

**Comment by logan-markewich** [comment]
I think this is good to merge, just need to resolve those conflicts 👍🏻

File: llama_index/vector_stores/tair.py (hunk 1 above)
Comment (seth-hg) [reply llama_index/vector_stores/tair.py:154]: > nit: could we use the utility in `llama_index/vector_stores/utils.py` to get metadata dict from the Node (and re-creating Node from metadata dict)?

node_to_metadata_dict() put all the extra key/value pairs in Node.node_info into a single json string. This makes it impossible for Tair to filter documents by these key/value pairs. Is that ok?

Comment (Disiok) [reply llama_index/vector_stores/tair.py:154]: Yes, this is actually intended. we use `extra_info` as the fields for metadata filtering, and `node_info` for other misc data of arbitrary types.

**Comment by Disiok** [comment]
Going to take over this PR, to make some small tweak on how metadata is treated and land it after!

**Pull request merged**